File: LeetCodeRecord/MethodFirst/543-DiameterOfBinaryTree.java
package MethodFirst;

class DiameterOfBinaryTree {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode (int x) {
            val = x;
        }
    }

    public int res = 0;

    public int diameterOfBinaryTree(TreeNode root) {

        getDepth(root);

        return res;
    }

    public int getDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = getDepth(root.left);
        int right = getDepth(root.right);

        if (this.res < left + right) {
            this.res = left + right;
        }

        return Math.max(left, right) + 1;
    }
}

File: Top200/543-DiameterOfBinaryTree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	_, dia := calcDiameter(root)

	return dia
}

func calcDiameter(root *TreeNode) (int, int) {
	if root == nil {
		return 0, 0
	}

	leftDep, leftDia := calcDiameter(root.Left)
	rightDep, rightDia := calcDiameter(root.Right)

	dia := leftDep + rightDep
	if leftDia > dia {
		dia = leftDia
	}
	if rightDia > dia {
		dia = rightDia
	}

	dep := leftDep
	if rightDep > dep {
		dep = rightDep
	}

	return dep + 1, dia
}
